controllers/backstage_controller.go

```diff
@@ -103,219 +104,140 @@ func (r *BackstageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	}
 
-	err := r.reconcileBackstageDeployment(ctx, &backstage, req.Namespace)
+	err = r.applyObjects(ctx, bsModel.Objects)
 	if err != nil {
-		return ctrl.Result{}, err
+		setStatusCondition(&backstage, bs.BackstageConditionTypeDeployed, metav1.ConditionFalse, bs.BackstageConditionReasonFailed, fmt.Sprintf("failed to apply backstage objects %s", err))
+		return ctrl.Result{}, fmt.Errorf("failed to apply backstage objects %w", err)
 	}
 
-	if err := r.reconcileBackstageService(ctx, &backstage, req.Namespace); err != nil {
-		return ctrl.Result{}, err
+	if err := r.cleanObjects(ctx, backstage); err != nil {
+		setStatusCondition(&backstage, bs.BackstageConditionTypeDeployed, metav1.ConditionFalse, bs.BackstageConditionReasonFailed, fmt.Sprintf("failed to clean backstage objects %s", err))
+		return ctrl.Result{}, fmt.Errorf("failed to clean backstage objects %w", err)
 	}
 
-	if r.IsOpenShift {
-		if err := r.reconcileBackstageRoute(ctx, &backstage, req.Namespace); err != nil {
-			return ctrl.Result{}, err
-		}
-	}
+	setStatusCondition(&backstage, bs.BackstageConditionTypeDeployed, metav1.ConditionTrue, bs.BackstageConditionReasonDeployed, "")
 
-	setStatusCondition(&backstage, bs.ConditionDeployed, v1.ConditionTrue, bs.DeployOK, "")
 	return ctrl.Result{}, nil
 }
 
-func (r *BackstageReconciler) readConfigMapOrDefault(ctx context.Context, name string, key string, ns string, object v1.Object) error {
+func (r *BackstageReconciler) applyObjects(ctx context.Context, objects []model.BackstageObject) error {
 
 	lg := log.FromContext(ctx)
 
-	if name == "" {
-		err := readYamlFile(defFile(key), object)
-		if err != nil {
-			return fmt.Errorf("failed to read YAML file: %w", err)
-		}
-		object.SetNamespace(ns)
-		return nil
-	}
+	for _, obj := range objects {
 
-	cm := corev1.ConfigMap{}
-	if err := r.Get(ctx, types.NamespacedName{Name: name, Namespace: ns}, &cm); err != nil {
-		return err
-	}
+		if err := r.Get(ctx, types.NamespacedName{Name: obj.Object().GetName(), Namespace: obj.Object().GetNamespace()}, obj.EmptyObject()); err != nil {
+			if !errors.IsNotFound(err) {
+				return fmt.Errorf("failed to get object: %w", err)
+			}
 
-	val, ok := cm.Data[key]
-	if !ok {
-		// key not found, default
-		lg.V(1).Info("custom configuration configMap and data exists, trying to apply it", "configMap", cm.Name, "key", key)
-		err := readYamlFile(defFile(key), object)
-		if err != nil {
-			return fmt.Errorf("failed to read YAML file: %w", err)
+			if err := r.Create(ctx, obj.Object()); err != nil {
+				return fmt.Errorf("failed to create object %w", err)
+			}
+
+			lg.V(1).Info("Create object ", "obj", obj.Object().GetName())
+			continue
 		}
-	} else {
-		lg.V(1).Info("custom configuration configMap exists but no such key, applying default config", "configMap", cm.Name, "key", key)
-		err := readYaml([]byte(val), object)
-		if err != nil {
-			return fmt.Errorf("failed to read YAML: %w", err)
+
+		if err := r.Update(ctx, obj.Object()); err != nil {
```

Comment:
Simply calling update() will cause unexpected result because the existing k8s object's meta data information (such as labels, annotations) may get lost. Administrators often may want to add additional labels or annotations in order to manager the cluster resources.
You should consider to use the CreateOrUpdate() library call used currently, or you may have to implement the logic to merge the configuration before calling update.

Reply:
As discussed, if we want Backstage owning the objects (which is the case) the only way to reshape runtime objects is to change default/raw/spec configuration, that is the desired state

Reply:
Yes, we do want to have runtime objects to be based on our configuration, however we should not prevent admins to add the labels/annotations to them, and this is often done as automation scripts. k8s may inject additional information to the metadata.  This will be a serious limitation which will not get accepted by the field.